File: CivitasRuby/lib/civitas/Casilla.rb
=begin
Authors: Esther García Gallego
         Yesenia González Dávila
         Grupo B3
=end
require_relative ./civitas/TipoCasilla.rb
class Casilla

  attr_reader :nombre, :tituloPropiedad

  @@carcel                #Atributo de instancia

  def self. constructorNombre (n)
    init()
    new(n, nil, 0, 0, nil)
  end

  def self. constructorTituloPropiedad (titulo)
    init()
    new(titulo.nombre, titulo, 0, 0, nil)
  end

  def self. constructorCantidadNombre (cantidad, n)
    init()
    new(n, nil, cantidad, 0, nil)
  end

  def self. constructorCarcelNombre (numCasillaCarcel, n)
    init()
    new(n, nil, 0, numCasillaCarcel, nil)
  end

  def self. constructorMazoNombre (n)
    init()
    new(n, nil, 0, 0, m)
  end

  def jugadorCorrecto(actual, todos)
  end

  def self. recibeJugador(actual, todos)
    #no entiendo qué hace este método
  end

  def toString
    str = ">> Casilla " + @nombre
    #Falta poner el resto de valores según el tipo de casilla (importe etc)
  end


  private #-------------------------------------------------------------------
  def initialize(n, titulo, cantidad, numCasillaCarcel, m)
    @nombre = n
    @importe = cantidad
    @@carcel = numCasillaCarcel
    @tituloPropiedad = titulo
    @tipo = Civitas::TipoCasilla::CALLE
    @mazo = m
  end

  def init
    @nombre = ""
    @importe = -1
    @@carcel = -1

    @tituloPropiedad = nil
    @sorpresa = nil
    @mazo = nil
  end

  def informe
    str = "El jugador " + todos[actual] + " ha caido en una casilla.\n" + toString()
    Diario.getInstance.ocurreEvento(str)
  end

  def recibeJugador_calle(actual, todos)
    if(jugadorCorrecto(actual, todos))
      informe(actual,todos)
      todos[actual].pagaAlquiler(tituloPropiedad.getPrecioAlquiler())
    end
  end

  def recibeJugador_impuesto(actual, todos)
    if (jugadorCorrecto(actual, todos))
      informe(actual, todos)
      todos[actual].pagaImpuesto(importe)
    end
  end

  def recibeJugador_juez(actual, todos)
    if(jugadorCorrecto(actual, todos))
      informe(actual, todos)
      todos[actual].encarcelar(carcel)
    end
  end

  def recibeJugador_sorpresa(actual, todos)
    if (jugadorCorrecto(actual, todos))
      informe(actual, todos)
      sorpresa.aplicarAJugador(actual, todos)
    end
  end
  
  def recibeJugador(iactual, todos)
    case @tipo
    when Civitas::TipoCasilla::CALLE
      recibeJugador_calle(iactual, todos)
    when Civitas::TipoCasilla::IMPUESTO
      recibeJugador_impuesto(iactual, todos)
    when Civitas::TipoCasilla::JUEZ
      recibeJugador_juez(iactual, todos)
    when Civitas::TipoCasilla::SORPRESA
      recibeJugador_sorpresa(iactual, todos)
    else
      informe(iactual, todos)
    end
    
  end
  
  
end

File: P5/civitasJava/src/civitas/Casilla.java
package civitas;
import java.util.ArrayList;

/**
 * @file Casilla.java
 * @author Yesenia González Dávila
 * @author Esther García Gallego
 * @class Casilla
 * @brief Crea las casillas del tablero.
 * @note Grupo B.3
*/

public class Casilla {

    /** Cambiamos las visibilidades a protected para que 
        la subclase pueda acceder a ellas. */
    protected String nombre;                      // Nombre casilla
    protected int carcel;                         // Juez
    protected float importe;                      // Impuesto
    protected Sorpresa sorpresa;                  // Sopresa
    protected MazoSorpresas mazo;                 // Mazo si la casilla es una sorpresa
    protected TituloPropiedad tituloPropiedad;    // Calle
    
    protected String tipoCasilla;                 // String para indicar el tipo de casilla
                                                  // (ahorrará redefinir el informe en todas las 
                                                  // subclases)
    
    /** Constructor descanso. */
    Casilla (String nombre) {
        mazo = null;
        tituloPropiedad = null;
        importe = -1f;
        carcel = -1;
        sorpresa = null;
        tipoCasilla = "DESCANSO";
        
        this.nombre = nombre;
    }

    /** Recibe a un jugador en la casilla. */
    void recibeJugador (int iactual, ArrayList<Jugador> todos) {
        informe (iactual, todos);
    }

    public String getNombre () {
        return nombre;
    }

    TituloPropiedad getTituloPropiedad() {
        return tituloPropiedad;
    }
    
    public boolean jugadorCorrecto (int actual, ArrayList<Jugador> todos) {
        return (actual < todos.size());
    }

    @Override
    public String toString() {
        String str = nombre ;
        return str;
    }

    protected void informe (int actual, ArrayList<Jugador> todos) {
        String str = "El jugador " + todos.get(actual).getNombre() + " cae en casilla " + tipoCasilla;
        Diario.getInstance().ocurreEvento(str);
    }
}